### PR TITLE
cache desc_tbl

### LINKED ISSUE #3622: Only first delivery of Fragment instance to BE carries TDescriptorTable cached for later re-use
## Enhancement
When FE delivers fragment instance to BE, the each rpc request carries the same,  entire TDescriptorTable, if  the query is complex, then both the size of TDescriptorTable and the number of fragment instances are very large, and the delivery is time-consuming.  so only first delivery of fragment instance to BE carries the TDescriptorTable, on receiving, BE cache it for later re-use, subsequential deliveries carry empty TDescriptorTable, in this way, the number of  bytes transfered can be reduced from size_of_TDescriptorTable * num_fragment_instances to size_of_TDescriptorTable * num_of_BEs.

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

https://github.com/StarRocks/starrocks/issues/3622

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 1]
```diff
@@ -473,79 +474,118 @@ public void exec() throws Exception {
             boolean isEnablePipelineEngine = ConnectContext.get() != null &&
                     ConnectContext.get().getSessionVariable().isEnablePipelineEngine() &&
                     (fragments.get(0).getSink() instanceof ResultSink);
+            boolean isSelect = isEnablePipelineEngine &&
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:477]: What‘s the meaning of `isSelect`?

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 2]
```diff
@@ -1944,12 +1987,13 @@ public FragmentExecParams(PlanFragment fragment) {
...
-            for (int i = 0; i < instanceExecParams.size(); ++i) {
-                final FInstanceExecParam instanceExecParam = instanceExecParams.get(i);
+            int startIdx = backendNum;
+            for (int i = 0; i < fInstanceExecParamList.size(); ++i) {
+                final FInstanceExecParam instanceExecParam = fInstanceExecParamList.get(i);
                 TExecPlanFragmentParams params = new TExecPlanFragmentParams();
 
                 if (exportSink != null && fileNamePrefix != null) {
-                    exportSink.setFileNamePrefix(fileNamePrefix + i + "_");
+                    exportSink.setFileNamePrefix(fileNamePrefix + (startIdx + i) + "_");
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1996]: why change this?

**Comment by stdpain** [comment]
How to ensure that the life cycle of DescriptorTbl is always valid.

**Comment by satanson** [comment]
> How to ensure that the life cycle of DescriptorTbl is always valid.

@stdpain 
DescriptorTbl is managed by QueryContext, QueryContext's lifetime contains fragment instances' lifetime.

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java (hunk 2 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1997]: in order to keep consistent with the original logic. In old implementation, toThrift is invoked only once for a PlanFragment, so for each FragmentInstance, here i increasing from zero is ok. but when desc_tbl cache is introduced, for a PlanFragment, toThrift is invoked twice, for the 2nd invocation, here i should increase from the base of 1st invocation.

**Review by stdpain — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Comment by kangkaisen** [comment]
LGTM，please merge it after daily test pass.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 3]
```diff
@@ -482,6 +484,13 @@
     @VariableMgr.VarAttr(name = SINGLE_NODE_EXEC_PLAN, flag = VariableMgr.INVISIBLE)
     private boolean singleNodeExecPlan = false;
 
+    @VariableMgr.VarAttr(name = ENABLE_DESC_TBL_CACHE)
+    private boolean enableDescTblCache = false;
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:488]: default false? Does user needs to understand this parameter?

Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:488]: I think user don't

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Review by stdpain — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by kangkaisen — CHANGES_REQUESTED** [review]
I think we needn't the `enableDescTblCache`  session  variable

**Review by stdpain — APPROVED** [review]

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**